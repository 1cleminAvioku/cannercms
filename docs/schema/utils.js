export const galleryValidation = {
  validator: (content, reject) => {
    if (content.length === 0) {
      return reject("should at least have one photo");
    }
  }
};