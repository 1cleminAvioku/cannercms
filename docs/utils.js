/* globals FIREBASE_API_KEY */

import {FirebaseClientService} from '@canner/image-service-config';
import firebase from 'firebase';
import {FirebaseRtdbClientConnector, PrismaClient} from 'canner-graphql-interface';


firebase.initializeApp({
  apiKey: FIREBASE_API_KEY,
  authDomain: "test-new-qa.firebaseapp.com",
  databaseURL: "https://test-new-qa.firebaseio.com",
  projectId: "test-new-qa",
  storageBucket: "test-new-qa.appspot.com",
  messagingSenderId: "983887338585"
});

const defaultApp = firebase.app();
const connector = new FirebaseRtdbClientConnector({
  database: defaultApp.database()
});


const storage = new FirebaseClientService({
  firebase: firebase,
  dir: "the/path/to", // specify the path you want upload to 
  filename: "filename", // rename file without extension
  hash: true, // if true, the filename will add a hash string, e.g.: `filename-${hash}.jpg`
}).getServiceConfig();

export default {
  storage,
  connector,
  graphClient: new PrismaClient()
}
