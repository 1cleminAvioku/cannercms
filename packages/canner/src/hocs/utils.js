// @flow
import {mapValues, get, isPlainObject, isArray} from 'lodash';

export function getValue(value: any, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (isPlainObject(result)) {
      if ('edges' in result && 'pageInfo' in result) {
        return get(result, ['edges', key, 'node']);
      }
      return get(result, key);
    } else if (isArray(result)) {
      return get(result, key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: any) {
  if (isPlainObject(value)) {
    if (value.edges && value.pageInfo) {
      return value.edges.map(edge => parseConnectionToNormal(edge.node));
    }
    return mapValues(value, item => parseConnectionToNormal(item));
  } else if (isArray(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

export function defaultValue(type: string, relation: any) {
  switch (type) {
    case 'connection': {
      return {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false
        }
      }
    }
    case 'array': {
      return [];
    }
    case 'object': {
      return {};
    }
    case 'boolean': {
      return false;
    }
    case 'number': {
      return 0;
    }
    case 'string': {
      return '';
    }
    case 'relation': {
      if (relation.type === 'toMany') {
        return {
          edges: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          }
        };
      } else {
        return null;
      }
    }
    case 'image':
    case 'file': {
      return {
        url: '',
        contentType: '',
        name: '',
        size: 0,
        __typename: null
      }
    }
    case 'geoPoint': {
      return {
        placeId: '',
        address: '',
        lat: 122,
        lng: 23
      };
    }
    default: {
      return null;
    }
  }
}