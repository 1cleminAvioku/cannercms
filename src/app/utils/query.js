/**
 * @flow
 */
type dataType = {
    totalPage: number,
    data: Array<any>
  } | Array<any>

import getByPath from 'lodash/get';

export default class Query {
  data: dataType

  constructor({data}: {data: Array<any>}) {
    this.data = data;
  }

  filter(filter: {[key: string]: any} = {}) {
    const data = this.getPureData();
    this.setData(Object.keys(filter).reduce((result, key) => {
      return this._filterByKey(result, key, filter[key]);
    }, data));
    return this;
  }

  sort(sort: {[key: string]: 1 | -1} = {}) {
    const data = this.getPureData();
    // eslint-disable-next-line max-len
    const collator = new Intl.Collator(undefined, {numeric: true, sensitivity: 'base', caseFirst: 'lower'});
    const key = Object.keys(sort)[0];
    const order = sort[key];
    this.setData(data.sort((a, b) => {
    // eslint-disable-next-line max-len
      return order > 0 ? collator.compare(a[key], b[key]) : collator.compare(b[key], a[key]);
    }));
    return this;
  }

  static turnPage(pagination?: {start: number, limit: number}) {
    if (!pagination) {
      return {page: 1, perPage: 10};
    }
    const {start, limit} = pagination;
    return {
      page: Math.ceil(start / limit) + 1,
      perPage: limit,
    };
  }

  page(pagination?: {page: number, perPage: number}) {
    const data = this.getPureData();
    if (!pagination) {
      return this;
    }
    const {page = 1, perPage = 10} = pagination;
    const queryData = data.slice(perPage * (page - 1), perPage * page);
    this.setData({
      goTo: (page) => ({
        start: (page - 1) * perPage,
        limit: perPage,
      }),
      page,
      totalPage: Math.ceil(data.length / perPage) || 0,
      data: queryData,
    });
    return this;
  }

  getPureData(): Array<any> {
    if (this.data && 'data' in this.data) {
      return (this.data: any).data;
    }
    return (this: any).data;
  }

  setData(data: dataType) {
    if (this.data && this.data.data) {
      (this.data: any).data = data;
    } else {
      this.data = data;
    }
  }

  getData() {
    return this.data;
  }

  // eslint-disable-next-line max-len
  _filterByKey(data: Array<any>, key: string, conditions: {[key: string]: any}) {
    return data.filter((datum) => {
      // eslint-disable-next-line max-len
      return Object.keys(conditions).reduce((result, conditionkey) => {
        // eslint-disable-next-line max-len
        return result && isValid(getByPath(datum, key.split('/'), ''), conditionkey, conditions[conditionkey]);
      }, true);
    });
  }
}

export function isValid(value: any, condition: string, conditionValue: any) {
  switch (condition) {
    case '$eq':
    case '$equal':
      // eslint-disable-next-line eqeqeq
      return value == conditionValue;
    case '$gt':
      return value > conditionValue;
    case '$gte':
      return value >= conditionValue;
    case '$lt':
      return value < conditionValue;
    case '$lte':
      return value <= conditionValue;
    case '$contains':
      return value.indexOf && value.indexOf(conditionValue) !== -1;
    case '$in':
      return conditionValue.indexOf && conditionValue.indexOf(value) !== -1;
    case '$regex':
      return new RegExp(conditionValue, 'g').test(value);
    default:
      return true;
  }
}

// eslint-disable-next-line max-len
export function passQuery(action: MutateAction, queryKey: string, totalPage: number) {
  if (action.type !== 'CREATE_ARRAY_ITEM') {
    return true;
  }
  // 用來判斷 CollectionUpadte 是否要在這個 query 的 data 下 mutate
  const allPage = queryKey === 'all';
  const query = allPage ? {filter: {}, pagination: {}} : JSON.parse(queryKey);
  const lastPage = getByPath(query, 'pagination.page', 1) === totalPage;
  const dataIsValid = new Query({data: [action.payload.value.toJS()]})
    .filter(query.filter).getPureData().length > 0;
  return dataIsValid && (allPage || lastPage);
}
