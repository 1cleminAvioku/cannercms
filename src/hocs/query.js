// @flow

import * as React from 'react';
import {Spin, Icon} from 'antd';
import type RefId from 'canner-ref-id';
import {Map, List, is} from 'immutable';
import Toolbar from './components/toolbar';
import type {Query} from '../query';
import {mapValues} from 'lodash';

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

type Props = {
  type: string,
  items: Object,
  refId: RefId,
  query: Query,
  fetch: FetchDef,
  subscribe: SubscribeDef,
  updateQuery: Function,
  ui: string,
  path: string,
  pattern: string,
  relation: {
    type: string,
    to: string
  },
  toolbar: {
    sort?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    pagination?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    filter?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    toolbarLayout?: {
      component?: React.ComponentType<*>,
      [string]: *
    }
  }
};

type State = {
  value: any,
  rootValue: any,
  originRootValue: any,
  isFetching: boolean,
}

export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<Props, State> {
    key: string;
    subscription: any;

    constructor(props: Props) {
      super(props);
      this.state = {
        value: null,
        rootValue: null,
        originRootValue: null,
        isFetching: true
      };
      this.key = props.refId.getPathArr()[0];
    }

    componentDidMount() {
      this.queryData();
    }

    UNSAFE_componentWillReceiveProps(props: Props) {
      const {refId} = this.props;
      
      if (refId.toString() !== props.refId.toString()) {
        this.queryData(props);
      } 
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    getRootValue = () => {
      return this.state.rootValue;
    }

    queryData = (props?: Props): Promise<*> => {
      const {refId, fetch} = props || this.props;
      return fetch(this.key).then(data => {
        this.setState({
          originRootValue: data,
          rootValue: parseConnectionToNormal(data),
          value: getValue(data, refId.getPathArr()),
          isFetching: false
        });
        this.subscribe();
      });
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = () => {
      const {subscribe, refId} = this.props;
      const subscription = subscribe(this.key, (newOriginRootValue) => {
        const {originRootValue} = this.state
        if (shouldUpdate(originRootValue, newOriginRootValue)) {
          this.setState({
            originRootValue: newOriginRootValue,
            rootValue: parseConnectionToNormal(newOriginRootValue),
            value: getValue(newOriginRootValue, refId.getPathArr()),
          });
        }
      });
      this.subscription = subscription;
    }

    render() {
      const {value, isFetching, rootValue} = this.state;
      const {toolbar, query, refId, items, type, path, relation, updateQuery, pattern} = this.props;
      if (isFetching) {
        return <Spin indicator={antIcon} />;
      }
      if (pattern === 'array' || (type === 'relation' && relation.type === 'toMany')) {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v]);
        return <Toolbar items={items} toolbar={toolbar} args={args} query={query} refId={refId} value={value} updateQuery={updateQuery}>
          <Com {...this.props} rootValue={rootValue} value={value.getIn(['edges'], new List()).map(item => item.get('node'))} />
        </Toolbar>;
      } else if (type === 'relation' && relation.type === 'toOne') {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v]);
        return <Toolbar items={items} toolbar={toolbar} args={args} query={query} refId={refId} value={value} updateQuery={updateQuery}>
          <Com {...this.props} rootValue={rootValue} value={value || new Map()} />
        </Toolbar>;
      }
      return <Com {...this.props} rootValue={rootValue} value={value} />;
    }
  };
}

export function getValue(value: Map<string, *>, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (Map.isMap(result)) {
      if (result.has('edges') && result.has('pageInfo')) {
        return result.getIn(['edges', key, 'node']);
      }
      return result.get(key);

    } else if (List.isList(result)) {
      return result.get(key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: Map<string, *> | List<*>) {
  if (Map.isMap(value)) {
    value = ((value: any): Map<string, any>);
    if (value.has('edges') && value.has('pageInfo')) {
      return (value.get('edges'): any).map(edge => parseConnectionToNormal(edge.get('node')));
    }
    return value.map(item => parseConnectionToNormal(item));
  } else if (List.isList(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

function shouldUpdate(value: any, newValue: any) {
  return !is(value, newValue);
}