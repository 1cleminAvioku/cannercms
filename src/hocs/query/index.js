// @flow

import * as React from 'react';
import Sort from './sort';
import Pagination from './pagination';
import Filter from './filter';
import isString from 'lodash/isString';
import isUndefined from 'lodash/isUndefined';
import type RefId from 'canner-ref-id';

type Props = {
  filter: filterType,
  sort: sortType,
  params: Object,
  routes: Array<string>,
  componentId: string,
  fetch: FetchDef,
  refId: RefId
}

type State = {
  filter: filterType,
  sort: sortType,
  pagination: {
    start: number,
    limit: number,
  },
  paginationRes: any,
  value: any,
}

export default (Plugin: React.ComponentType<*>) => {
  return class PluginWithQuery extends React.PureComponent<Props, State> {
    constructor(props: Props) {
      super(props);
      this.state = {
        filter: {},
        sort: {},
        pagination: {
          start: 0,
          limit: 10,
        },
        paginationRes: {},
        value: null,
      };
      this.changeFilter = this.changeFilter.bind(this);
      this.changeSort = this.changeSort.bind(this);
      this.changePagination = this.changePagination.bind(this);
    }

    componentWillMount() {
      const {refId, subscribe, componentId} = this.props;
      this.queryData()
        .then(() => {
          subscribe(refId.toString(), componentId, 'value', (value) => {
            this.setState({
              value,
            });
          });
        });
    }

    queryData() {
      const {refId, fetch, componentId} = this.props;
      const {filter, sort, pagination} = this.state;
      return fetch(refId.toString(), componentId, {filter, sort, pagination}).then((ctx) => {
        this.setState({
          paginationRes: ctx.response.pagination,
          value: ctx.response.body,
        });
      });
    }

    changeFilter(condition) {
      this.setState({
        filter: condition,
      }, this.queryData);
    }

    changeSort(condition) {
      this.setState({
        sort: condition,
      }, this.queryData);
    }

    changePagination(start, limit) {
      this.setState({
        pagination: {
          start,
          limit,
        },
      }, this.queryData);
    }

    render() {
      const {filter, sort, params, routes} = this.props;
      const {paginationRes, pagination, value} = this.state;
      const op = params && params.op;
      if (!value) {
        return null;
      }
      return (
        <div>
          {
            // 如果是 string 則不是交給這個 filter 處理
            // 是 restful-qa 會在 render 前 filter
            // 目前只有 ireconcile 使用到
            op === 'create' || routes.length > 1 || isString(filter) || isUndefined(filter) ?
              null :
              <Filter onChange={this.changeFilter} schema={filter}/>
          }
          {
            op === 'create' || routes.length > 1 || isUndefined(sort) ?
              null :
              <Sort onChange={this.changeSort} options={sort}/>
          }
          <Plugin {...this.props} query={{filter: this.state.filter, sort: this.state.sort, pagination}} rootValue={value} value={value} showPagination={false}/>
          {
            op === 'create' || routes.length > 1 ?
              null :
              <Pagination onChange={this.changePagination} pagination={pagination} paginationRes={paginationRes}/>
          }
        </div>
      );
    }
  };
};

