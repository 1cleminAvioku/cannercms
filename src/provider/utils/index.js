// @flow
export {default as createEmptyData} from './createEmptyData';
export {default as generateId} from './generateId';
export {default as transformData} from './transformData';
