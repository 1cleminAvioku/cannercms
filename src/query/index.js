// @flow

import {schemaToQueriesObject, objectToQueries} from './utils';
import {get, set} from 'lodash';

export class Query {
  schema: Object
  queries: Object

  constructor({schema}: {schema: Object}) {
    this.schema = schema;
    this.queries = schemaToQueriesObject(schema);
  }

  updateQueries = (pathArr: Array<string>, field: string, value: any) => {
    const path = `${pathArr.join('.fields.')}.${field}`;
    set(this.queries, path, value);
  }

  getQueries = (pathArr: Array<string>): Object => {
    const path = pathArr.join('.fields.');
    return get(this.queries, path);
  }

  toGQL = (pathArr: Array<string>): string => {
    const obj = this.getQueries(pathArr);
    return objectToQueries(obj);
  }
}