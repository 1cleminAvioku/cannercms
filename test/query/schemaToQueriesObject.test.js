import {schemaToQueriesObject} from '../../src/query/utils';

describe('schema to queries object', () => {
  let schema, queriesObject;
  beforeEach(() => {
    schema = {
      posts: {
        keyName: 'posts',
        type: 'array',
        items: {
          type: 'object',
          items: {
            title: {
              keyName: 'title',
              type: 'string',
            },
            staredPosts: {
              keyName: 'staredPosts',
              type: 'relation',
              relation: {
                type: 'toMany',
                to: 'posts'
              }
            },
            bestAuthor: {
              keyName: 'bestAuthor',
              type: 'relation',
              relation: {
                type: 'toOne',
                to: 'users',
                typename: 'User'
              }
            }
          }
        }
      },
      users: {
        type: 'array',
        items: {
          type: 'object',
          items: {
            name: {
              type: 'string'
            },
            email: {
              type: 'string'
            },
            images: {
              type: 'array',
              items: {
                type: 'image'
              }
            },
            posts: {
              type: 'relation',
              relation: {
                type: 'toMany',
                to: 'posts'
              }
            }
          }
        }
      },
    };
  })

  it('should works', () => {
    expect(schemaToQueriesObject(schema)).toMatchSnapshot();
  });
});