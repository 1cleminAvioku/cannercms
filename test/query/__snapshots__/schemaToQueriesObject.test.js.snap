// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema to queries object should works 1`] = `
Object {
  "posts": Object {
    "args": Object {
      "pagination": Object {
        "first": 10,
      },
    },
    "fields": Object {
      "bestAuthor": Object {
        "args": Object {
          "pagination": Object {
            "first": 10,
          },
        },
        "fields": Object {
          "email": null,
          "id": null,
          "images": null,
          "name": null,
          "posts": null,
        },
      },
      "id": null,
      "staredPosts": Object {
        "args": Object {
          "pagination": Object {
            "first": 10,
          },
        },
        "fields": Object {
          "bestAuthor": null,
          "id": null,
          "staredPosts": null,
          "title": null,
        },
        "isPlural": true,
      },
      "title": null,
    },
    "isPlural": true,
  },
  "users": Object {
    "args": Object {
      "pagination": Object {
        "first": 10,
      },
    },
    "fields": Object {
      "email": null,
      "id": null,
      "images": null,
      "name": null,
      "posts": Object {
        "args": Object {
          "pagination": Object {
            "first": 10,
          },
        },
        "fields": Object {
          "bestAuthor": null,
          "id": null,
          "staredPosts": null,
          "title": null,
        },
        "isPlural": true,
      },
    },
    "isPlural": true,
  },
}
`;
